### ISSUE TITLE
[DN-DAB-DETR] The output of ONNX's Mul OP is different from the TFLite's output.

### ISSUE THREAD

**Issue by On-JungWoan** [PINTO0309/onnx2tf#327]
### Issue Type

Others

### onnx2tf version number

1.9.12

### onnx version number

1.13.1

### onnxruntime version number

1.13.1

### onnxsim (onnx_simplifier) version number

0.4.17

### tensorflow version number

2.12.0

### Download URL for ONNX

https://drive.google.com/file/d/1rjqhNfn85we2IG6YwKOJlhKv-fMRFha7/view?usp=sharing

### Parameter Replacement JSON

```json
{
    "format_version": 1,
    "operations": [


      {
        "op_name": "/backbone/backbone.1/Unsqueeze",
        "param_target": "outputs",
        "param_name": "/backbone/backbone.1/Unsqueeze_output_0",
        "post_process_transpose_perm": [3,0,1,2]
      },
      {
        "op_name": "/backbone/backbone.1/Unsqueeze_1",
        "param_target": "outputs",
        "param_name": "/backbone/backbone.1/Unsqueeze_1_output_0",
        "post_process_transpose_perm": [3,0,1,2]
      }

    ]      
  }
```

### Description

Hi! @PINTO0309. I'm trying to convert DN-DAB-DETR to TFLite. However, when I multiply two tensors, the output of ONNX is different from the TFLite's output. Below is my result using `-cotof`.

- `-cotof` **result**

![image](https://user-images.githubusercontent.com/84084372/234170811-59d0008b-a65c-47f0-9dea-14cfa145b7b2.png)

- **my onnx graph**

![image](https://user-images.githubusercontent.com/84084372/234169949-3486dd4b-cec2-4015-9b4d-018aa0111c02.png)

- **python script**

```python
test = pos*pos_scales
```

Should I avoid tensor's multiply?

**Comment by PINTO0309** [comment]
Thank you.

I will be very busy until the second week of May and will not have enough time to debug and maintain at the same pace as before.

![image](https://user-images.githubusercontent.com/33194443/234430882-6004f1b2-daa3-4acd-949f-e48a9165e111.png)

If the accuracy check is `Unmatched` on this simple model, then the problem appears to be in the accuracy check logic itself. But I don't have time to debug.

**Comment by PINTO0309** [comment]
- https://github.com/PINTO0309/onnx2tf/releases/tag/1.9.13
- `onnx2tf.py`
  - Added sanitizing of `:` and `/` just before first dummy inference
  - [ScaleMul.onnx.zip](https://github.com/PINTO0309/onnx2tf/files/11328382/ScaleMul.onnx.zip)
  - Before
    ```
    onnx2tf \
    -i ScaleMul.onnx \
    -kat onnx__Mul_0 onnx__Mul_1 \
    -cind "onnx::Mul_0" pos.npy \
    -cind "onnx::Mul_1" pos_scales.npy \
    -cotof
    ```
  - After
    ```
    onnx2tf \
    -i ScaleMul.onnx \
    -kat onnx__Mul_0 onnx__Mul_1 \
    -cind "onnx__Mul_0" pos.npy \
    -cind "onnx__Mul_1" pos_scales.npy \
    -cotof
    ```

- onnx
  ![image](https://user-images.githubusercontent.com/33194443/234450238-1983fb76-838f-43c7-b2aa-4e2a565e1c70.png)
- result
  ![image](https://user-images.githubusercontent.com/33194443/234450028-bf949636-87ef-4e6f-ac18-cfcf8e121868.png)
- tflite
  ![image](https://user-images.githubusercontent.com/33194443/234450360-ae654db7-7be9-4f8b-b89c-59f7a1902d53.png)

**Comment by On-JungWoan** [comment]
Thank you for your detail answer. I get the `Matched` message in small model. However, I still get the `Unmatched` message in original model.

```
onnx2tf -i dn_dab_detr_480x480.onnx.onnx  -prf replace.json \
-kat input.1 -onimc /transformer/encoder/Mul_output_0
```

![image](https://user-images.githubusercontent.com/84084372/234496098-7b3554b6-aff0-4660-9c24-9b99cc5cb80f.png)

<br>
<br>

While I'm trying to find the reasons of the problem, I noticed strange things. The original model(tflite) was only taking one input for the Mul OP instead of two. The other inputs appeared to be simply stored as a binary file. Is this the reason for this problem?

- **Big model**

    - [Download tflite link](https://drive.google.com/file/d/1bIIVC01kKHA-BazIC14BFIBDudr4qAIi/view?usp=sharing)

<p align="center"><img src="https://user-images.githubusercontent.com/84084372/234498860-164d8f0b-760c-4a8c-8ea6-0bddf6f69087.png"></p>

- **Small model**

<p align="center"><img src="https://user-images.githubusercontent.com/84084372/234492820-4789ff95-497a-4704-a6b4-1b4b8eda9b7d.png"></p>

**Comment by PINTO0309** [comment]
> While I'm trying to find the reasons of the problem, I noticed strange things. The original model(tflite) was only taking one input for the Mul OP instead of two.

To me this is not a strange phenomenon at all.

![image](https://user-images.githubusercontent.com/33194443/234565310-adea46b8-aac6-4674-bdde-40a3a6ae0ce0.png)

![image](https://user-images.githubusercontent.com/33194443/234565625-5f8d1c3c-743e-4677-bee0-ba44bd5636be.png)

![image](https://user-images.githubusercontent.com/33194443/234565838-12ba5110-0107-485f-ba1a-347dd57dccb1.png)

**Comment by PINTO0309** [comment]
The value is already broken in `/backbone/backbone.0/Gather`.

```
INFO: onnx_op_type: Resize onnx_op_name: /backbone/backbone.0/Resize
INFO:  input_name.1: /backbone/backbone.0/Constant_output_0 shape: [1, 1, 480, 480] dtype: <class 'numpy.float32'>
INFO:  input_name.2:  shape: None dtype: None
INFO:  input_name.3:  shape: None dtype: None
INFO:  input_name.4: /backbone/backbone.0/Concat_1_output_0 shape: [4] dtype: <class 'numpy.int64'>
INFO:  output_name.1: /backbone/backbone.0/Resize_output_0 shape: [1, 1, 15, 15] dtype: float32
INFO: tf_op_type: upsampling2d_nearest
INFO:  input.1.images: shape: (1, 480, 480, 1) dtype: float32 
INFO:  input.2.boxes: 
INFO:  input.3.box_indices: 
INFO:  input.4.new_size/crop_size: shape: (2,) dtype: <dtype: 'int32'> 
INFO:  input.5.method: val: nearest 
INFO:  input.6.extrapolation_value: val: 0.0 
INFO:  input.7.align_corners: val: False 
INFO:  output.1.output: shape: (1, 15, 15, 1) dtype: <dtype: 'float32'> 

INFO: onnx_op_type: Gather onnx_op_name: /backbone/backbone.0/Gather
INFO:  input_name.1: /backbone/backbone.0/Resize_output_0 shape: [1, 1, 15, 15] dtype: float32
INFO:  input_name.2: /Constant_output_0 shape: [] dtype: <class 'numpy.int64'>
INFO:  output_name.1: /backbone/backbone.0/Gather_output_0 shape: [1, 15, 15] dtype: float32
INFO: tf_op_type: gather_v2
INFO:  input.1.params: shape: (1, 15, 15, 1) dtype: <dtype: 'float32'> 
INFO:  input.2.indices: shape: () dtype: <dtype: 'int64'> 
INFO:  input.3.axis: val: 0 
INFO:  output.1.output: shape: (15, 15, 1) dtype: <dtype: 'float32'>

INFO: onnx_op_type: CumSum onnx_op_name: /backbone/backbone.1/CumSum
INFO:  input_name.1: /backbone/backbone.1/Cast_output_0 shape: [1, 15, 15] dtype: float32
INFO:  input_name.2: /backbone/backbone.1/Constant_1_output_0 shape: [] dtype: <class 'numpy.int32'>
INFO:  output_name.1: /backbone/backbone.1/CumSum_output_0 shape: [1, 15, 15] dtype: float32
INFO: tf_op_type: cumsum
INFO:  input.1.x: shape: (15, 15, 1) dtype: <dtype: 'float32'> 
INFO:  input.2.axis: shape: () dtype: int32 
INFO:  input.3.exclusive: val: False 
INFO:  input.4.reverse: val: False 
INFO:  output.1.output: shape: (15, 15, 1) dtype: <dtype: 'float32'> 

INFO: onnx_op_type: CumSum onnx_op_name: /backbone/backbone.1/CumSum_1
INFO:  input_name.1: /backbone/backbone.1/Cast_output_0 shape: [1, 15, 15] dtype: float32
INFO:  input_name.2: /backbone/backbone.1/Constant_2_output_0 shape: [] dtype: <class 'numpy.int32'>
INFO:  output_name.1: /backbone/backbone.1/CumSum_1_output_0 shape: [1, 15, 15] dtype: float32
INFO: tf_op_type: cumsum
INFO:  input.1.x: shape: (15, 15, 1) dtype: <dtype: 'float32'> 
INFO:  input.2.axis: shape: () dtype: int32 
INFO:  input.3.exclusive: val: False 
INFO:  input.4.reverse: val: False 
INFO:  output.1.output: shape: (15, 15, 1) dtype: <dtype: 'float32'> 
```

**Comment by On-JungWoan** [comment]
Thank you for kindly responding despite your busy schedule. If it's not too much trouble, could you tell me how to print the log above? I will try to find the problem myself so as not to bother you.

**Comment by PINTO0309** [comment]
Note that only the `Unsqueeze` transposition specification has a special specification.

https://github.com/PINTO0309/onnx2tf#parameter-replacement

```json
{
  "format_version": 1,
  "operations": [

    {
      "op_name": "/backbone/backbone.0/Gather",
      "param_target": "attributes",
      "param_name": "axis",
      "values": 3
    },
    {
      "op_name": "/backbone/backbone.1/Unsqueeze",
      "param_target": "op",
      "new_shape": [1,15,15,1]
    },
    {
      "op_name": "/backbone/backbone.1/Unsqueeze_1",
      "param_target": "op",
      "new_shape": [1,15,15,1]
    }

  ]
}
```

**Comment by On-JungWoan** [comment]
I made modifications to the model as follows and finally got `matched` message in all the outputs. As you mentioned, there seems to be a bug in the process of the `Mul` operator handling constants.

- **Before**
<p align="center"><img src="https://user-images.githubusercontent.com/84084372/234635922-9aa7ae12-cd27-43e4-b603-02233ab926e8.png"></p>

<p align="center"><img src="https://user-images.githubusercontent.com/84084372/234638730-9d093760-7ab4-4ef4-857d-5c97738a7957.png"></p>

- **After**


<p align="center"><img src="https://user-images.githubusercontent.com/84084372/234636039-87b6af44-8e62-4c59-b6e9-90ef084fdb68.png"></p>

<p align="center"><img src="https://user-images.githubusercontent.com/84084372/234638829-0caf5b70-45bd-476f-b310-ef0a258c891f.png"></p>

- **Result**

<p align="center"><img src="https://user-images.githubusercontent.com/84084372/234636178-926f6f36-aaee-448a-aa35-7a087006c5b6.png"></p>

**Comment by On-JungWoan** [comment]
By the way, there seems to be some error in the Placeholder operator here and there, is it okay to ignore them?

![image](https://user-images.githubusercontent.com/84084372/234640931-3b6284af-8972-494d-b2b8-49d88c29680d.png)

**Comment by On-JungWoan** [comment]
Thank you for your fast response. I will check the issue that you mentioned. Can I close this issue?

**Comment by PINTO0309** [comment]
It is up to you to decide whether to close the issue.

**Comment by On-JungWoan** [comment]
I am currently having a problem with the Gemm operator. All other outputs are correct, but only the output of the Gemm operator shows an `Unmatched` message. Is this just an error in the calculation method?

![image](https://user-images.githubusercontent.com/84084372/234777443-e4bb1a85-f648-43f0-a1aa-f5a51564d61e.png)

![image](https://user-images.githubusercontent.com/84084372/234779205-354d7e29-eef0-4731-9090-14dabdeb1f9f.png)


**ONNX Link**
- <https://drive.google.com/file/d/1Nh6xRv__1USQoBMCeK61fHhCAgxbsotN/view?usp=sharing>


<details>
<summary><strong>replace.json</strong></summary>

```
{
  "format_version": 1,
  "operations": [
    {
      "op_name": "/backbone/backbone.1/Unsqueeze",
      "param_target": "op",
      "new_shape": [1,15,15,1]
    },
    {
      "op_name": "/backbone/backbone.1/Unsqueeze_1",
      "param_target": "op",
      "new_shape": [1,15,15,1]
    }
  ]
}
```

</details>

<br>
<br>
<br>

The strange thing here is that when I checked the `-cotof` result of the operator using the `-kat` option, I was able to get a Matched message.



```shell
onnx2tf -i new_dn_dab_detr_400x400.onnx \
-prf replace.json \
-onimc /transformer/encoder/layers.0/self_attn/Gemm_output_0 \
-cotof -cotoa 1e-1
```

![image](https://user-images.githubusercontent.com/84084372/234777882-b41f7f5f-eaf4-4d9c-aa83-87638b8c72ac.png)

Should I modify the replace.json file?

**Comment by PINTO0309** [comment]
The ONNX file you shared with me was corrupted and unreadable when I downloaded it.

**Comment by On-JungWoan** [comment]
Oh, I'm sorry. That's my mistake. Here is the new ONNX file link.

- **ONNX Link**
    - <https://drive.google.com/file/d/17kpDkkXegRBieuufQLWhZeeCV5aYRsBy/view?usp=sharing>

**Comment by PINTO0309** [comment]
>The strange thing here is that when I checked the -cotof result of the operator using the -kat option, I was able to get a Matched message.

Please give me as much info as possible. 

```
onnx2tf -i new_dn_dab_detr_480x480.onnx \
-prf replace.json \
-onimc /transformer/encoder/layers.0/self_attn/Gemm_output_0 \
-cotof -cotoa 1e-1
```
This command you are posting does not include the `-kat` option and is unable to reproduce normal and abnormal patterns, respectively.

**Comment by On-JungWoan** [comment]
# 1. Model shown in image

> Also, how did you generate the model for this image? If I can't reproduce the model you have on my PC, the man-hours required for testing will be very large, and it will be very difficult for me.

First, the part shown in the image is the MultiHeadAttention process, and below is the link to the corresponding model.

- ONNX link
    - <https://drive.google.com/file/d/1Ojk0mqoUCdYDsL9mKPFWKvrnkWGwGM4K/view?usp=sharing>

<br>

For this part of model, `matched` messages are displayed for all outputs.

```
onnx2tf -i multiheadatn.onnx \
-kat key_padding_mask \
-cotof -cotoa 1e-1
```

![image](https://user-images.githubusercontent.com/84084372/234903125-5f75a6db-434d-4742-9c74-f31c625a1196.png)

<br>
<br>

# 2. The part I think is weird

## 2-1. `-onimic` for the first layer's Gemm_output

The `matched` message is displayed for all outputs even when `-onimic` is applied to the `Gemm` output of the first layer(`/transformer/encoder/layers.0/self_attn/Gemm_output_0`).

```
onnx2tf -i new_dn_dab_detr_480x480.onnx \
-prf replace.json \
-onimc /transformer/encoder/layers.0/self_attn/Gemm_output_0 \
-cotof -cotoa 1e-1 \
-kat input.1
```

![image](https://user-images.githubusercontent.com/84084372/234899296-ca790ddf-f66b-4edd-bbe7-7a1c81552fc3.png)

![image](https://user-images.githubusercontent.com/84084372/234899842-bf2fdbb0-8789-4aea-a77c-c5ad95f3ce07.png)

<br>

## 2-2. `-onimic` for the second layer's Gemm_output

However, when I apply `-onimic` to the second layer(`/transformer/encoder/layers.1/self_attn/Gemm_output_0`), the `Gemm` output of the first layer, which was correct until now, becomes incorrect.

```
onnx2tf -i new_dn_dab_detr_480x480.onnx \
-prf replace.json \
-onimc /transformer/encoder/layers.1/self_attn/Gemm_output_0 \
-cotof -cotoa 1e-1 \
-kat input.1
```
![image](https://user-images.githubusercontent.com/84084372/234900158-4d0432de-cafc-4e38-af20-18c8f4fc5388.png)

**Comment by On-JungWoan** [comment]
I found a very strange issue while experimenting with nn.multiheadattention. When inputting a value that exceeds a certain threshold into the TFLite converted multiheadattention, the results of ONNX and TFLite differ. I have been struggling with this issue all weekend, and I think this is a clear bug. I'm sorry to bother you when you're busy, but I would really appreciate it if you could take a look.

- **.ipynb script**
    <https://drive.google.com/file/d/1YjkDZO_dHTPvrgQED2t15FjF_MBiZFtp/view?usp=sharing>

- **Max value < 0.01**

```python
scale=1e-2

( ... )

print(len(res[res==False]))
```
```
0
```

- **Max value < 10**

```python
scale=1e+1

( ... )

print(len(res[res==False]))
```
```
150092
```

<br>
<br>

### Model

- **Download link**

    <https://drive.google.com/drive/folders/1g5-D246ydFKqTmQP8ZSVTfcPUJGdoPuL?usp=sharing>


- **Pytorch Model**

```python
class MHAtn(nn.Module):
    def __init__(self):
        super().__init__()
        self.self_attn = nn.MultiheadAttention(256, 8, 0.0)
    def forward(self, q, k, src):
        return self.self_attn(q, k, src)
```

**Comment by PINTO0309** [comment]
Thank you. I am allocating some time to implement the research, so I will be a little slow in responding for a while.

**Comment by On-JungWoan** [comment]
I thought this issue was related to Matmul, so I read the issue you wrote below.

- [#317](https://github.com/PINTO0309/onnx2tf/issues/317)

> Frequent accuracy degradation occurs only when the sizes of all dimensions except the batch size are the same, as in [1,256,256] in the figure below.

- [#133](https://github.com/PINTO0309/onnx2tf/issues/133)

> This means that for tensors of structures other than 4 or 5 dimensional, the tool may make a wrong decision about transposition.

<br>
<br>

On the other hand, wouldn't it be possible to adjust the dimension of the weight that is passed to the input of `Matmul` to 4-dimensions for `nn.Linear`?

![image](https://user-images.githubusercontent.com/84084372/235420581-61681a4e-40e9-4bae-ade9-b41fb622bd85.png)

If the dimension of the weight in `nn.Linear` is `255,255`, is there any solution to this?

**Comment by PINTO0309** [comment]
WIP

- colabo
  https://drive.google.com/file/d/1YjkDZO_dHTPvrgQED2t15FjF_MBiZFtp/view?usp=sharing

  ```
  onnx2tf -i multiheadatn.onnx -kat query key value
  ```

- onnx
  [multiheadatn.onnx.zip](https://github.com/PINTO0309/onnx2tf/files/11364981/multiheadatn.onnx.zip)
- tflite
  [multiheadatn_float32.tflite.zip](https://github.com/PINTO0309/onnx2tf/files/11364982/multiheadatn_float32.tflite.zip)

**Comment by PINTO0309** [comment]
We already know that the `256x256` `MatMul` pattern makes transposition errors and causes errors, but before that, there appears to be a fatal problem with TFLite's runtime.

**Comment by On-JungWoan** [comment]
You mean, this issue is related to a third party?

**Comment by PINTO0309** [comment]
Yes, I am. I think the results speak for themselves. However, I feel that a little more research should be done to determine when the phenomenon occurs and at what threshold the problem occurs.

**Comment by On-JungWoan** [comment]
Thank you for your response. I will also continue to research to solve the problem.

**Comment by AlexanderLutsenko** [comment]
> I am currently having a problem with the Gemm operator. All other outputs are correct, but only the output of the Gemm operator shows an `Unmatched` message. Is this just an error in the calculation method?

Hello! As you want to convert a pytorch model to tensorflow and and only need onnx as an intermediary, [Nobuco converter](https://github.com/AlexanderLutsenko/nobuco) might be of help. It was specifically designed with ease of debugging in mind. The way it handles mismatching tensor layouts is quite neat, you should never encounter errors related to wrong channel dimension.
For example,  `MHAtn` module can be converted in this fashion:

```python
import os
os.environ["CUDA_VISIBLE_DEVICES"] = ""

import nobuco
from nobuco import ChannelOrder, ChannelOrderingStrategy
from nobuco.layers.weight import WeightLayer

class MHAtn(nn.Module):
    def __init__(self):
        super().__init__()
        self.self_attn = nn.MultiheadAttention(256, 8, 0.0)

    def forward(self, q, k, src, key_padding_mask):
        return self.self_attn(q, k, src, key_padding_mask)


L = 256
N = 1
q = torch.normal(0, 1, size=(L, N, 256))
k = torch.normal(0, 1, size=(L, N, 256))
v = torch.normal(0, 1, size=(L, N, 256))
m = torch.normal(0, 1, size=(N, L))

pytorch_module = MHAtn().eval()

keras_model = nobuco.pytorch_to_keras(
    pytorch_module,
    args=[q, k, v, m],
    inputs_channel_order=ChannelOrder.PYTORCH,
    outputs_channel_order=ChannelOrder.PYTORCH,
)
```

**Comment by On-JungWoan** [comment]
@AlexanderLutsenko Thank for your help. But your converter is not supported `torch.nn.functional._canonical_mask`.

**Comment by AlexanderLutsenko** [comment]
> @AlexanderLutsenko Thank for your help. But your converter is not supported `torch.nn.functional._canonical_mask`.
@On-JungWoan shouldn't be a problem to add support for it. What does the module look like in pytorch?

**Comment by On-JungWoan** [comment]
Below is the script of `F._canonical_mask`.

```python
def _canonical_mask(
        mask: Optional[Tensor],
        mask_name: str,
        other_type: Optional[DType],
        other_name: str,
        target_type: DType,
        check_other: bool = True,
) -> Optional[Tensor]:

    if mask is not None:
        _mask_dtype = mask.dtype
        _mask_is_float = torch.is_floating_point(mask)
        if _mask_dtype != torch.bool and not _mask_is_float:
            raise AssertionError(
                f"only bool and floating types of {mask_name} are supported")
        if check_other and other_type is not None:
            if _mask_dtype != other_type:
                warnings.warn(
                    f"Support for mismatched {mask_name} and {other_name} "
                    "is deprecated. Use same type for both instead."
                )
        if not _mask_is_float:
            mask = (
                torch.zeros_like(mask, dtype=target_type)
                .masked_fill_(mask, float("-inf"))
            )
    return mask

```

**Comment by PINTO0309** [comment]
```
Name: torch
Version: 2.0.0
Summary: Tensors and Dynamic neural networks in Python with strong GPU acceleration
Home-page: https://pytorch.org/
Author: PyTorch Team
Author-email: packages@pytorch.org
License: BSD-3
Location: /usr/local/lib/python3.8/dist-packages
Requires: filelock, jinja2, networkx, nvidia-cublas-cu11, nvidia-cuda-cupti-cu11, nvidia-cuda-nvrtc-cu11, nvidia-cuda-runtime-cu11, nvidia-cudnn-cu11, nvidia-cufft-cu11, nvidia-curand-cu11, nvidia-cusolver-cu11, nvidia-cusparse-cu11, nvidia-nccl-cu11, nvidia-nvtx-cu11, sympy, triton, typing-extensions
Required-by: ci-sdr, complexPyTorch, cvpods, efficientnet-pytorch, espnet, facexlib, fairseq, g-mlp-pytorch, geffnet, gfpgan, ibug-face-detection, lpips, mish-cuda, openpifpaf, pretrainedmodels, pytorch-ignite, pytorch-lightning, pytorch-lightning-bolts, pytorch-metric-learning, pytorch-msssim, pytorch-wavelets, realesrgan, smplx, spatial-correlation-sampler, thop, timm, torchaudio, torchdata, torchgeometry, torchmetrics, torchtext, torchvision, triton

```

**Comment by PINTO0309** [comment]
torch==1.13.1 did indeed convert successfully. :+1: 

![image](https://user-images.githubusercontent.com/33194443/235459691-b57d9c8e-b15b-4439-a788-d9a03d7cf76f.png)

**Comment by PINTO0309** [comment]
Thanks.

Assuming the generated tflite is correct with nobco's and onnx2tf, I assume the inference results will be broken. nobco's tflite will take a little longer to test because it does not have the signature embedded.

Also, the weights of `MatMul` are different from @On-JungWoan's, so I'm having trouble comparing the exact inference results.

@On-JungWoan, can I ask you to do a comparison? I don't have your PyTorch source code so I can't test it accurately.

**Comment by AlexanderLutsenko** [comment]
> I am currently having a problem with the Gemm operator. All other outputs are correct, but only the output of the Gemm operator shows an `Unmatched` message. Is this just an error in the calculation method?
> 
> ![image](https://user-images.githubusercontent.com/84084372/234777443-e4bb1a85-f648-43f0-a1aa-f5a51564d61e.png)
> 
> ![image](https://user-images.githubusercontent.com/84084372/234779205-354d7e29-eef0-4731-9090-14dabdeb1f9f.png)
> 
> **ONNX Link**
> 
>     * https://drive.google.com/file/d/1Nh6xRv__1USQoBMCeK61fHhCAgxbsotN/view?usp=sharing
> 
> 
> **replace.json**
> 
> 
> 
> The strange thing here is that when I checked the `-cotof` result of the operator using the `-kat` option, I was able to get a Matched message.
> 
> ```shell
> onnx2tf -i new_dn_dab_detr_480x480.onnx \
> -prf replace.json \
> -onimc /transformer/encoder/layers.0/self_attn/Gemm_output_0 \
> -cotof -cotoa 1e-1
> ```
> 
> ![image](https://user-images.githubusercontent.com/84084372/234777882-b41f7f5f-eaf4-4d9c-aa83-87638b8c72ac.png)
> 
> Should I modify the replace.json file?

@On-JungWoan BTW, how's there `MatMul` op in you .tflite file? When I call `TFLiteConverter.from_keras_model_file`, it gets split into multiple `FullyConnected` layers in parallel.

**Comment by PINTO0309** [comment]
@AlexanderLutsenko

https://github.com/PINTO0309/onnx2tf/blob/890794138faccce504fc8360e68d718ad197aafc/onnx2tf/onnx2tf.py#L963

**Comment by On-JungWoan** [comment]
For now, it seems that there are a lot of operators that are not supported by `nobuco` in the 2.0.0 version of torch. Therefore, I am also having difficulty making an accurate comparison.

![image](https://user-images.githubusercontent.com/84084372/235466370-e743917c-e8c8-4af0-b63d-b33ffa024ff9.png)

- **Error log**

```
 │   │   │  TransformerDecoderLayer[DN_DETR.models.DN_DAB_DETR.transformer](float32_106<300,1,256>, float32_36<625,1,256>, float32_37<625,1,256>, float32_124<300,1,256>, float32_125<300,1,256>, False) -> float32_129<300,1,256>
 │   │   │   │   I  File "/data/ojw/convert/DN_DETR/models/DN_DAB_DETR/transformer.py", line 556 
 │   │   │   │   D  File "/home/user/anaconda3/envs/tflite/lib/python3.9/site-packages/torch/nn/modules/activation.py", line 1249  
 │   │   │   │  PReLU[torch.nn.modules.activation](float32_126<300,1,2048>) -> float32_128<300,1,2048>
 │   │   │   │   │   I  File "/home/user/anaconda3/envs/tflite/lib/python3.9/site-packages/torch/nn/modules/activation.py", line 1309 
 │   │   │   └   └· prelu[torch.nn.functional](float32_126<300,1,2048>, float32_127<1>) -> float32_128<300,1,2048>
 │   │   │   I  File "/data/ojw/convert/DN_DETR/models/DN_DAB_DETR/transformer.py", line 51 
 │   │   │  requires_grad_[torch.Tensor](float32_130<128>, True) -> float32_130<128>
 │   │   │   I  File "/data/ojw/convert/DN_DETR/models/DN_DAB_DETR/transformer.py", line 58 
 │   │   │  sin[torch.Tensor](float32_131<300,1,64>) -> float32_132<300,1,64>
 │   │   │   I  File "/data/ojw/convert/DN_DETR/models/DN_DAB_DETR/transformer.py", line 58 
 │   │   │  cos[torch.Tensor](float32_133<300,1,64>) -> float32_134<300,1,64>
 │   │   │   I  File "/data/ojw/convert/DN_DETR/models/DN_DAB_DETR/transformer.py", line 59 
 │   │   │  sin[torch.Tensor](float32_135<300,1,64>) -> float32_136<300,1,64>
 │   │   │   I  File "/data/ojw/convert/DN_DETR/models/DN_DAB_DETR/transformer.py", line 59 
 │   │   │  cos[torch.Tensor](float32_137<300,1,64>) -> float32_138<300,1,64>
 │   │   │   I  File "/data/ojw/convert/DN_DETR/models/DN_DAB_DETR/transformer.py", line 65 
 │   │   │  sin[torch.Tensor](float32_139<300,1,64>) -> float32_140<300,1,64>
 │   │   │   I  File "/data/ojw/convert/DN_DETR/models/DN_DAB_DETR/transformer.py", line 65 
 │   │   │  cos[torch.Tensor](float32_141<300,1,64>) -> float32_142<300,1,64>
 │   │   │   I  File "/data/ojw/convert/DN_DETR/models/DN_DAB_DETR/transformer.py", line 69 
 │   │   │  sin[torch.Tensor](float32_143<300,1,64>) -> float32_144<300,1,64>
 │   │   │   I  File "/data/ojw/convert/DN_DETR/models/DN_DAB_DETR/transformer.py", line 69 
 │   │   │  cos[torch.Tensor](float32_145<300,1,64>) -> float32_146<300,1,64>
 │   │   │   I  File "/data/ojw/convert/DN_DETR/models/DN_DAB_DETR/transformer.py", line 303 
 │   │   │   D  File "/data/ojw/convert/DN_DETR/models/DN_DAB_DETR/transformer.py", line 443  
 │   │   │  TransformerDecoderLayer[DN_DETR.models.DN_DAB_DETR.transformer](float32_129<300,1,256>, float32_36<625,1,256>, float32_37<625,1,256>, float32_147<300,1,256>, float32_148<300,1,256>, False) -> float32_152<300,1,256>
 │   │   │   │   I  File "/data/ojw/convert/DN_DETR/models/DN_DAB_DETR/transformer.py", line 556 
 │   │   │   │   D  File "/home/user/anaconda3/envs/tflite/lib/python3.9/site-packages/torch/nn/modules/activation.py", line 1249  
 │   │   │   │  PReLU[torch.nn.modules.activation](float32_149<300,1,2048>) -> float32_151<300,1,2048>
 │   │   │   │   │   I  File "/home/user/anaconda3/envs/tflite/lib/python3.9/site-packages/torch/nn/modules/activation.py", line 1309 
 │   │   │   └   └· prelu[torch.nn.functional](float32_149<300,1,2048>, float32_150<1>) -> float32_151<300,1,2048>
 │   │   │   I  File "/data/ojw/convert/DN_DETR/models/DN_DAB_DETR/transformer.py", line 51 
 │   │   │  requires_grad_[torch.Tensor](float32_153<128>, True) -> float32_153<128>
 │   │   │   I  File "/data/ojw/convert/DN_DETR/models/DN_DAB_DETR/transformer.py", line 58 
 │   │   │  sin[torch.Tensor](float32_154<300,1,64>) -> float32_155<300,1,64>
 │   │   │   I  File "/data/ojw/convert/DN_DETR/models/DN_DAB_DETR/transformer.py", line 58 
 │   │   │  cos[torch.Tensor](float32_156<300,1,64>) -> float32_157<300,1,64>
 │   │   │   I  File "/data/ojw/convert/DN_DETR/models/DN_DAB_DETR/transformer.py", line 59 
 │   │   │  sin[torch.Tensor](float32_158<300,1,64>) -> float32_159<300,1,64>
 │   │   │   I  File "/data/ojw/convert/DN_DETR/models/DN_DAB_DETR/transformer.py", line 59 
 │   │   │  cos[torch.Tensor](float32_160<300,1,64>) -> float32_161<300,1,64>
 │   │   │   I  File "/data/ojw/convert/DN_DETR/models/DN_DAB_DETR/transformer.py", line 65 
 │   │   │  sin[torch.Tensor](float32_162<300,1,64>) -> float32_163<300,1,64>
 │   │   │   I  File "/data/ojw/convert/DN_DETR/models/DN_DAB_DETR/transformer.py", line 65 
 │   │   │  cos[torch.Tensor](float32_164<300,1,64>) -> float32_165<300,1,64>
 │   │   │   I  File "/data/ojw/convert/DN_DETR/models/DN_DAB_DETR/transformer.py", line 69 
 │   │   │  sin[torch.Tensor](float32_166<300,1,64>) -> float32_167<300,1,64>
 │   │   │   I  File "/data/ojw/convert/DN_DETR/models/DN_DAB_DETR/transformer.py", line 69 
 │   │   │  cos[torch.Tensor](float32_168<300,1,64>) -> float32_169<300,1,64>
 │   │   │   I  File "/data/ojw/convert/DN_DETR/models/DN_DAB_DETR/transformer.py", line 303 
 │   │   │   D  File "/data/ojw/convert/DN_DETR/models/DN_DAB_DETR/transformer.py", line 443  
 │   │   │  TransformerDecoderLayer[DN_DETR.models.DN_DAB_DETR.transformer](float32_152<300,1,256>, float32_36<625,1,256>, float32_37<625,1,256>, float32_170<300,1,256>, float32_171<300,1,256>, False) -> float32_175<300,1,256>
 │   │   │   │   I  File "/data/ojw/convert/DN_DETR/models/DN_DAB_DETR/transformer.py", line 556 
 │   │   │   │   D  File "/home/user/anaconda3/envs/tflite/lib/python3.9/site-packages/torch/nn/modules/activation.py", line 1249  
 │   │   │   │  PReLU[torch.nn.modules.activation](float32_172<300,1,2048>) -> float32_174<300,1,2048>
 │   │   │   │   │   I  File "/home/user/anaconda3/envs/tflite/lib/python3.9/site-packages/torch/nn/modules/activation.py", line 1309 
 └   └   └   └   └· prelu[torch.nn.functional](float32_172<300,1,2048>, float32_173<1>) -> float32_174<300,1,2048>
```

**Comment by On-JungWoan** [comment]
@AlexanderLutsenko Below is `MultiHeadAttn` in tflite.

![image](https://user-images.githubusercontent.com/84084372/235467002-7baf7c35-4b6a-4bfe-869e-aba2956c630c.png)

**Comment by AlexanderLutsenko** [comment]
@On-JungWoan If sin, cos, requires_grad and prelu is all there is, these can be implemented in no time. Also, why not just convert and compare `nn.MultiheadAttention` in isolation?

**Comment by AlexanderLutsenko** [comment]
> @AlexanderLutsenko Below is `MultiHeadAttn` in tflite.
> 
> ![image](https://user-images.githubusercontent.com/84084372/235467002-7baf7c35-4b6a-4bfe-869e-aba2956c630c.png)

@On-JungWoan Beautiful, how'd you get that?

**Comment by On-JungWoan** [comment]
@AlexanderLutsenko This is a tflite converted by `onnx2tf` created by @PINTO0309.

**Comment by AlexanderLutsenko** [comment]
@On-JungWoan And why the hell can't google's own tool do the same?

**Comment by PINTO0309** [comment]
I think nobuco's conversion result is correct. I have never seen a complete tool that can generate TensorFlow models directly from PyTorch, so I am very excited.

I'm leaving this issue for a while to move on to other work, but I'll have to look at the TFLite runtime implementation to see if the inference results are broken.

Thanks, @AlexanderLutsenko @On-JungWoan

**Comment by On-JungWoan** [comment]
@AlexanderLutsenko I was really surprised. The output that didn't match before now matches perfectly. It's really amazing.

```python
res = np.isclose(
    tf_out,
    torch_out
)
len(res[res==False])
```

```
0
```

**Comment by On-JungWoan** [comment]
@AlexanderLutsenko However, most of the functions in torch.nn.functional (such as sin, cos, etc.) are implemented in C or C++. Is it possible to implement these operators as well?

**Comment by PINTO0309** [comment]
The output now matches even onnx2tf. As I thought, it was a bug in the TFLite runtime. 

There seems to be an overflow problem with `BatchMatMul`. Thus, I found that this issue can be solved by just specifying the `-ebu` option to get the same results as nobuco.

https://colab.research.google.com/drive/13ujjRdftMFnX1yvxy4EdyxROHVAQLKTe?usp=sharing

- `scale = 1e+1`
![image](https://user-images.githubusercontent.com/33194443/235474729-5c78032b-ebc5-4eb0-a9e6-31a4b8fe9416.png)

**Comment by PINTO0309** [comment]
@On-JungWoan issue seems to be resolved so I will close this issue. 

@AlexanderLutsenko It was very helpful. Thank you very much.

**Comment by On-JungWoan** [comment]
Thank you so much @PINTO0309, @AlexanderLutsenko !

**Comment by On-JungWoan** [comment]
Hi! @PINTO0309. Thank you so much for your hard work. I finally get the `Matched` message from the `MultiHeadAttention` OPs. But, In whole model, I still get the `Unmatched` message in the mostly OPs. Is it better to reopen the another issue?

**Comment by PINTO0309** [comment]
@On-JungWoan 

Could you please share your modified ONNX file in the comments below? The meaningless `Resize` adjustment is quite cumbersome.
https://github.com/PINTO0309/onnx2tf/issues/327#issuecomment-1523703643

**Comment by On-JungWoan** [comment]
Hi! @PINTO0309. Thank you for your hard work. Actually, there were some additional modifications made after the ONNX model mentioned in <https://github.com/PINTO0309/onnx2tf/issues/327#issuecomment-1523703643>. Below is the link to the modified ONNX model.

- ONNX & Input file

    - https://drive.google.com/drive/folders/1N8zCPifA-IZJtP2Hh-ccNCScmYL-Ywdw?usp=sharing

<br>

Additionally, when testing the ONNX model, I use the following command.

```
onnx2tf -i mod_dn_dab_detr.onnx -cotof -cotoa 1e-1 -kat input.1 -cind input.1 input.npy
```

**Comment by PINTO0309** [comment]
The output is now completely normal. Hmmm. What did you do?

```
onnx2tf -i mod_dn_dab_detr.onnx -cotof 
```

![image](https://user-images.githubusercontent.com/33194443/235816564-322bb423-d9e1-44aa-b3da-b65c4d3f0a3f.png)

**Comment by PINTO0309** [comment]
Oh, I'm sorry. I forgot to enter the test data. I deleted the comment I posted just before because it was an error.

**Comment by PINTO0309** [comment]
I even identified the problem here. Test data is not properly transposed from NHWC to NCHW.

https://github.com/PINTO0309/onnx2tf/blob/b1cf6e78c2f1ab278545def27ccbd168c8dd3cb8/onnx2tf/utils/common_functions.py#L3522-L3529

**Comment by PINTO0309** [comment]
Win. :+1: 

![image](https://user-images.githubusercontent.com/33194443/236079342-45e9890e-23b4-4676-af6f-e9c58716ae37.png)

**Comment by PINTO0309** [comment]
Please wait a little longer. Right now I am in the middle of a regression test with my CI. I will upgrade to v1.10.0 when it is all green.

**Comment by On-JungWoan** [comment]
I understand it. Thank you so much.

**Comment by PINTO0309** [comment]
https://github.com/PINTO0309/onnx2tf/releases/tag/1.10.0

The following commands can be used for conversion. Please try it.

Originally, the problem was quite complicated by the combination of multiple problems with the way the `-cind` option handled the test data and the `Softmax` and `MatMul` dimensional correction process.

  ```
  onnx2tf -i mod_dn_dab_detr.onnx -cotof
  ```
  ![image](https://user-images.githubusercontent.com/33194443/236079054-e43cc81e-bfca-419b-9910-a82ef9fdbea5.png)

**Comment by On-JungWoan** [comment]
Thank you for your hard work. I finally get the `Matched` message in the result of the `-cotof` option, but when I inference the converted tflite model, I still get the `-inf` values in half of the output. Is this bug in TFLite runtime? 

<br>

- **My result file** (Changed input shape from `1,3,800,800` to `1,3,800,801`)

    - <https://drive.google.com/drive/folders/18gnXX8P8dB_1LaglvnZw1ibbfD7H3wAj?usp=sharing>

- **Inference script**

```python
import numpy as np
import tensorflow as tf

TFLITE_PATH = 'my/tflite/path/final_model_float32.tflite'

# Load the TFLite model and allocate tensors.
interpreter = tf.lite.Interpreter(model_path=TFLITE_PATH)
interpreter.allocate_tensors()
# Get input and output tensors.
input_details = interpreter.get_input_details()
output_details = interpreter.get_output_details()

interpreter.set_tensor(
    input_details[0]['index'],
    np.ones(list(input_details[0]['shape'])).astype(np.float32)
)

interpreter.invoke()

out_list = []
for i in range(len(output_details)):
    out_list.append(interpreter.get_tensor(output_details[i]['index']))

print(out_list[0])
```
```
array([[[-inf, -inf, -inf, ..., -inf, -inf, -inf],
         [-inf, -inf, -inf, ..., -inf, -inf, -inf],
         [-inf, -inf, -inf, ..., -inf, -inf, -inf],
         ...,
         [-inf, -inf, -inf, ..., -inf, -inf, -inf],
         [-inf, -inf, -inf, ..., -inf, -inf, -inf],
         [-inf, -inf, -inf, ..., -inf, -inf, -inf]]], dtype=float32)
```

**Comment by On-JungWoan** [comment]
I still get `Unmatched` message in the result of the `-cotof` option.

- **Command**

```
onnx2tf -i mod_dn_dab_detr.onnx -cotof
```
```
Unmatched
```

<br>

```
onnx2tf -i final_model.onnx -cotof
```
```
Unmatched
```

- **Versions**

```
Name: onnx2tf
Version: 1.10.0
Summary: Self-Created Tools to convert ONNX files (NCHW) to TensorFlow/TFLite/Keras format (NHWC). The purpose of this tool is to solve the massive Transpose extrapolation problem in onnx-tensorflow (onnx-tf).
Home-page: https://github.com/PINTO0309/onnx2tf
Author: Katsuya Hyodo
Author-email: rmsdh122@yahoo.co.jp
License: MIT License
Location: /home/user/anaconda3/envs/tflite/lib/python3.9/site-packages
Requires: 
Required-by: simple-onnx-processing-tools
---
Name: onnxsim
Version: 0.4.17
Summary: Simplify your ONNX model
Home-page: https://github.com/daquexian/onnx-simplifier
Author: ONNX Simplifier Authors
Author-email: daquexian566@gmail.com
License: Apache License v2.0
Location: /home/user/anaconda3/envs/tflite/lib/python3.9/site-packages
Requires: onnx, rich
Required-by: 
---
Name: onnx
Version: 1.13.1
Summary: Open Neural Network Exchange
Home-page: https://github.com/onnx/onnx
Author: ONNX
Author-email: onnx-technical-discuss@lists.lfaidata.foundation
License: Apache License v2.0
Location: /home/user/anaconda3/envs/tflite/lib/python3.9/site-packages
Requires: numpy, protobuf, typing-extensions
Required-by: onnx-graphsurgeon, onnxsim
---
Name: onnxruntime-gpu
Version: 1.13.1
Summary: ONNX Runtime is a runtime accelerator for Machine Learning models
Home-page: https://onnxruntime.ai
Author: Microsoft Corporation
Author-email: onnxruntime@microsoft.com
License: MIT License
Location: /home/user/anaconda3/envs/tflite/lib/python3.9/site-packages
Requires: coloredlogs, flatbuffers, numpy, packaging, protobuf, sympy
Required-by: 
---
Name: tensorflow
Version: 2.12.0rc0
Summary: TensorFlow is an open source machine learning framework for everyone.
Home-page: https://www.tensorflow.org/
Author: Google Inc.
Author-email: packages@tensorflow.org
License: Apache 2.0
Location: /home/user/anaconda3/envs/tflite/lib/python3.9/site-packages
Requires: absl-py, astunparse, flatbuffers, gast, google-pasta, grpcio, h5py, jax, keras, libclang, numpy, opt-einsum, packaging, protobuf, setuptools, six, tensorboard, tensorflow-estimator, tensorflow-io-gcs-filesystem, termcolor, typing-extensions, wrapt
Required-by: 

```

**Comment by PINTO0309** [comment]
The ONNX file you shared with me is corrupt and cannot be displayed on Netron.

> I still get the -inf values in half of the output. Is this bug in TFLite runtime?

If the accuracy check after conversion is normal, I can only surmise that either a problem on the runtime side or the fact that all input data is 1 could be the cause of the problem.

**Comment by On-JungWoan** [comment]
I will use your docker envirionments. Thank you for your answer.

**Comment by On-JungWoan** [comment]
Always thank you for your hard work. I tested in the docker environment you shared and was able to confirm that the matched message is displayed in the `-cotof` results(in `mod_dn_dab_detr.onnx`). However, when I actually perform inference with tflite, the output is different from onnx. In most cases with `float16`, the output values are different with onnx, and even with `float32`, half of the values come out as `-inf`. I got these results even when I performed inference in your docker environment. How can I solve this problem?

![image](https://user-images.githubusercontent.com/84084372/236266060-d9bbe94c-68fc-4ecf-b896-6e9b258070c5.png)

```python
import os
import pickle
import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)
warnings.simplefilter(action='ignore', category=Warning)
warnings.simplefilter(action='ignore', category=DeprecationWarning)
warnings.simplefilter(action='ignore', category=FutureWarning)
warnings.simplefilter(action='ignore', category=RuntimeWarning)
import random
random.seed(0)
import numpy as np
np.set_printoptions(
    precision=6,
    floatmode='fixed',
    suppress=True,
    edgeitems=3,
    linewidth=100,
)
np.random.seed(0)

os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'
import tensorflow as tf
from tensorflow.lite.python import interpreter as iw

TFLITE_PATH = 'saved_model/mod_dn_dab_detr_float32.tflite'
interpreter = iw.Interpreter(
    model_path=TFLITE_PATH,
    num_threads=4,
)
input_details = interpreter.get_input_details()
input_shape_1 = input_details[0]['shape']
output_details = interpreter.get_output_details()

test_data1 = np.random.randn(*input_shape_1).astype(np.float32)

interpreter.allocate_tensors()
interpreter.set_tensor(
    input_details[0]['index'],
    test_data1,
)
interpreter.invoke()
output_data = interpreter.get_tensor(output_details[0]['index'])

print('')
print('tflite ===============================================')
print(output_data)


# with open('output/tflite_out.pkl', 'wb') as f:
#     pickle.dump(out_list, f)
```

**Comment by PINTO0309** [comment]
I can't say whether the following is really the correct test data to input into this model, because I don't know. Is it correct to enter random values? Since the accuracy check of the model is normal, it is better to suspect first the parts of the model other than its structure.

```python
test_data1 = np.random.randn(*input_shape_1).astype(np.float32)
```

You may want to find out which operations have divergent values.

**Comment by On-JungWoan** [comment]
I have tried various methods other than the test dummy data mentioned below, but I always got a value of `-inf`. Btw, the strange thing here is that only `mod_dn_dab_detr_float32` outputs a value of `-inf`. In `mod_dn_dab_detr_float32`, although it is not quite right, it does produce a value in float16.

```python
np.ones(input_shape_1).astype(np.float32) # unmatched
torch.rand(input_shape_1, dtype=torch.float32).numpy() # unmatched
np.random.rand(1,800,800,3).astype(np.float32) # unmatched

# etc ...
```

**Comment by PINTO0309** [comment]
https://s3.ap-northeast-2.wasabisys.com/temp-models/onnx2tf_327/test/mod_dn_dab_detr.onnx
https://s3.ap-northeast-2.wasabisys.com/temp-models/onnx2tf_327/test/mod_dn_dab_detr_float32.tflite
https://s3.ap-northeast-2.wasabisys.com/temp-models/onnx2tf_327/test/mod_dn_dab_detr_float16.tflite

I cannot reproduce it.

```
onnx2tf -i mod_dn_dab_detr.onnx -coion
```

```python
import os
import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)
warnings.simplefilter(action='ignore', category=Warning)
warnings.simplefilter(action='ignore', category=DeprecationWarning)
warnings.simplefilter(action='ignore', category=FutureWarning)
warnings.simplefilter(action='ignore', category=RuntimeWarning)
import random
random.seed(0)
import numpy as np
np.set_printoptions(
    precision=6,
    floatmode='fixed',
    suppress=True,
    edgeitems=3,
    linewidth=100,
)
np.random.seed(0)
os.environ['CUDA_VISIBLE_DEVICES'] = '-1'
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'
import tensorflow as tf
import onnxruntime

ONNX_PATH = 'mod_dn_dab_detr.onnx'
ort_session = onnxruntime.InferenceSession(ONNX_PATH, providers=['CPUExecutionProvider'])
input_name = ort_session.get_inputs()[0].name
input_shape = ort_session.get_inputs()[0].shape
output_num = len(ort_session.get_outputs())

test_data = np.ones(input_shape, dtype=np.float32)

ort_inputs = {
    ort_session.get_inputs()[0].name: test_data,
}
onnx_out = ort_session.run(
    ['7370'],
    ort_inputs,
)

TFLITE_PATH = 'mod_dn_dab_detr_float32.tflite'
interpreter = tf.lite.Interpreter(model_path=TFLITE_PATH)
tf_lite_model = interpreter.get_signature_runner()
tflite_output_float32 = tf_lite_model(
    input_1=test_data.transpose(0,2,3,1),
)
TFLITE_PATH = 'mod_dn_dab_detr_float16.tflite'
interpreter = tf.lite.Interpreter(model_path=TFLITE_PATH)
tf_lite_model = interpreter.get_signature_runner()
tflite_output_float16 = tf_lite_model(
    input_1=test_data.transpose(0,2,3,1),
)

print('=====================================================')
print(onnx_out[0].shape)
print(tflite_output_float32['7370'].shape)
print(tflite_output_float16['7370'].shape)

print('onnx =================================================')
print(onnx_out[0])
print('')
print('tflite float32 =======================================')
print(tflite_output_float32['7370'])
print('')
print('tflite float16 =======================================')
print(tflite_output_float32['7370'])
```
```
=====================================================
(1, 300, 60)
(1, 300, 60)
(1, 300, 60)
onnx =================================================
[[[ -9.889225  -5.058238  -4.400069 ...  -4.518981  -4.422979  -5.619113]
  [ -9.666203  -5.503964  -4.429813 ...  -4.284597  -3.964971  -4.635274]
  [-10.165474  -5.403321  -4.708091 ...  -4.778224  -4.726847  -5.837096]
  ...
  [-10.237301  -5.250232  -4.427892 ...  -4.601456  -4.263485  -5.674369]
  [ -9.821136  -5.886144  -4.720225 ...  -4.816403  -4.191678  -4.233029]
  [ -9.900963  -5.236600  -4.576357 ...  -4.596010  -4.396852  -5.339815]]]

tflite float32 =======================================
[[[ -9.889221  -5.058239  -4.400070 ...  -4.518981  -4.422979  -5.619112]
  [ -9.666204  -5.503964  -4.429811 ...  -4.284597  -3.964969  -4.635273]
  [-10.165472  -5.403322  -4.708091 ...  -4.778225  -4.726850  -5.837097]
  ...
  [-10.237301  -5.250231  -4.427890 ...  -4.601455  -4.263482  -5.674368]
  [ -9.821136  -5.886144  -4.720224 ...  -4.816402  -4.191678  -4.233030]
  [ -9.900966  -5.236598  -4.576355 ...  -4.596011  -4.396853  -5.339818]]]

tflite float16 =======================================
[[[ -9.889221  -5.058239  -4.400070 ...  -4.518981  -4.422979  -5.619112]
  [ -9.666204  -5.503964  -4.429811 ...  -4.284597  -3.964969  -4.635273]
  [-10.165472  -5.403322  -4.708091 ...  -4.778225  -4.726850  -5.837097]
  ...
  [-10.237301  -5.250231  -4.427890 ...  -4.601455  -4.263482  -5.674368]
  [ -9.821136  -5.886144  -4.720224 ...  -4.816402  -4.191678  -4.233030]
  [ -9.900966  -5.236598  -4.576355 ...  -4.596011  -4.396853  -5.339818]]]
```

**Comment by On-JungWoan** [comment]
> tflite_output_float32 = tf_lite_model(
    input_1=test_data.transpose(0,2,3,1),
)

I'm so sorry to bothering you. But, how can I get the input name?

```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/unist/anaconda3/envs/onx_test/lib/python3.9/site-packages/tensorflow/lite/python/interpreter.py", line 237, in __call__
    raise ValueError('Invalid Input name (%s) for SignatureDef' %
ValueError: Invalid Input name (input_1) for SignatureDef
```

**Comment by PINTO0309** [comment]
```python
import os
import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)
warnings.simplefilter(action='ignore', category=Warning)
warnings.simplefilter(action='ignore', category=DeprecationWarning)
warnings.simplefilter(action='ignore', category=FutureWarning)
warnings.simplefilter(action='ignore', category=RuntimeWarning)
import random
random.seed(0)
import numpy as np
np.set_printoptions(
    precision=6,
    floatmode='fixed',
    suppress=True,
    edgeitems=3,
    linewidth=100,
)
np.random.seed(0)
os.environ['CUDA_VISIBLE_DEVICES'] = '-1'
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'
import tensorflow as tf
import onnxruntime

ONNX_PATH = 'mod_dn_dab_detr.onnx'
ort_session = onnxruntime.InferenceSession(ONNX_PATH, providers=['CPUExecutionProvider'])
input_name = ort_session.get_inputs()[0].name
input_shape = ort_session.get_inputs()[0].shape
output_num = len(ort_session.get_outputs())

test_data = np.ones(input_shape, dtype=np.float32)

ort_inputs = {
    ort_session.get_inputs()[0].name: test_data,
}
onnx_out = ort_session.run(
    ['7370'],
    ort_inputs,
)

TFLITE_PATH = 'mod_dn_dab_detr_float32.tflite'
interpreter = tf.lite.Interpreter(model_path=TFLITE_PATH)
tf_lite_model = interpreter.get_signature_runner()
inputs = {'input.1': test_data.transpose(0,2,3,1)}
tflite_output_float32 = tf_lite_model(
    **inputs
)
TFLITE_PATH = 'mod_dn_dab_detr_float16.tflite'
interpreter = tf.lite.Interpreter(model_path=TFLITE_PATH)
tf_lite_model = interpreter.get_signature_runner()
inputs = {'input.1': test_data.transpose(0,2,3,1)}
tflite_output_float16 = tf_lite_model(
    **inputs
)

print('=====================================================')
print(onnx_out[0].shape)
print(tflite_output_float32['7370'].shape)
print(tflite_output_float16['7370'].shape)

print('onnx =================================================')
print(onnx_out[0])
print('')
print('tflite float32 =======================================')
print(tflite_output_float32['7370'])
print('')
print('tflite float16 =======================================')
print(tflite_output_float32['7370'])
```

**Comment by On-JungWoan** [comment]
Your code works very well. I want to express my gratitude for always helping me. As a token of my appreciation, I have decided to become your sponsor. Although I may not be able to provide a large amount of support as a student, I still want to show my gratitude to you in this way.

**Comment by PINTO0309** [comment]
> Although I may not be able to provide a large amount of support as a student, I still want to show my gratitude to you in this way.

You don't have to strain yourself. I earned all of my tuition money on my own and attended college. I am well aware that there are times in our lives when we struggle financially. 

When you graduate and earn a lot of money, please buy me a beer.

**Comment by On-JungWoan** [comment]
Thank you for saying so. I will definitely do that 😊
